bpo-30095: Customizing HTMLCalendar

oz123 opened the pull request:
as discussed here http://bugs.python.org/issue30095

Unfortunately, this whole module isn't pep8 compatible, I didn't touch where I am not supposed to touch, so pep8 won't passed here.

oz123 committed:
Add more styling options to HTMLCalendar

The HTMLCalendar has now multiple class attributes for styling the
calendar.
In order to retain backwards compatibility style names are not changed.

The new attributes are:

  weekday_head_styles
  month_head_styles = ("month",)
  year_head_styles = ("year",)
  year_styles = ("year",)

This are added to the previously existing styles, which where scattered
around the code hard coded or as class attributes:

  month style
  day styles
  noday style

The relevant methods have been modified to use these attributes. If you
want a calendar where elements has multiple CSS attributes you can
now do the following:

>>> my_styles = HTMLCalendar.cssclasses[:]
>>> my_stelys = " ".join((s, "black text-centered")) for s in my_styles]

class MyCalendar(HTMLCalendar):
   cssclasses = styles
   month_head_styles = "month-title text-bold"

This implementation simply replaces hard coded string with class
attributes (which are also single strings albite the attribute names which are
in plural).

oz123 committed:
Consitent naming

oz123 committed:
Add test case for subclassing

oz123 committed:
Add documentation for customizing HTMLCalendar

oz123 committed:
Documentation updates

 Fix typos, and more class attributes documentation

oz123 commented:
@doerwalter, I updated the documentation. I would be happy to hear your comments now.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
       sheet should be used. *encoding* specifies the encoding to be used for the
       output (defaulting to the system default encoding).
 
+   :class:`HTMLCalendar` has the following attribute you can override to customize the style of your calender:
```
attribute -> attributes
Try not to use 'you' or 'your'.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
+
+  .. attribute:: cssclasses
+
+    Is a list of CSS styles used for each weekday. The default style list is::
```
Not sure this should start with 'is'.  The calendar attributes on this page all start with a subject, such as 'An array'.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
+
+        cssclasses = ["mon text-bold", "tue", "wed", "thu", "fri", "sat", "sun red"]
+
+    Note, that the length of this list must be 7 items.
```
I don't think this needs a comma after Note.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
+
+  .. attribute:: noday_classes
+
+     The CSS class for a week day occuring in previous or the coming month.
```
- occuring -> occurring.
- week day -> weekday
- "in the previous or coming month"

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
+
+     Is a list of CSS styles used for each weekday. The default style list is
+
+     the same as ``HTMLCalendar.cssclassess``.
```
- Not sure about starting with 'is'.  
- These two lines gave odd spacing when I looked at the html rendered.
- There's an extra 's' at the end of HTMLCalendar.cssclasses.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
+
+  .. attribute:: month_head_classes
+
+     A space separted list of styles for the month head, for example::
```
separted -> separated

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
+
+  .. attribute:: month_classes
+
+     a space separted list of styles for the whole table.
```
- Capitalize 'a'.
- separted -> separated
- Also, there is a blank line in between the description and the default value for this, year_classes and year_head_classes.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
+  .. attribute:: year_classes
+
+     a space separted list of styles for the table when formatting the year
+     as a table of tables (see: ``HTMLCalendar.formatyear``).
```
- Capitalize 'a'.
- separted -> separated

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,67 @@ it's the base calendar for all computations.
+
+  .. attribute:: year_head_classes
+
+     a space separted list of styles for the table head when formatting the year as a table of tables (see: ``HTMLCalendar.formatyear``).
```
- capitalize 'a'
- separted -> separated

csabella commented:
When I looked at the html page for the docs, the indentation was off compared to the section before your changes (I used 'make html' to be able to view it).  I'm also wondering about the name of 'cssclasses'.  Since the year and month ones have year and month in the name, maybe this should too?

oz123 commented:
@csabella Thanks for the thorough review. I am embarrassed there are so many spelling errors...

I would prefer to rename the ``cssclasses`` attribute to ``weekday_classes``, but I was afraid somebody is already using it with that name.
If there is no objection to that, I would rename that attribute. I will correct the spelling errors as soon as possible.

csabella commented:
@oz123  Please don't be embarrassed about it.  I probably should have just asked you to spellcheck it instead of commenting on each one.  Sorry about that.

As far as the attribute name, maybe @doerwalter could help resolve that.  I just found it inconsistent when reading the doc, so I wanted to ask about it.

oz123 committed:
Correct spelling and fix styling

 Thanks @csabllea

oz123 commented:
@doerwalter what do you think about renaming the attribute HTMLCalendar.cssclasses` to  `HTMLCalendar.weekday_styles`?

csabella commented:
@oz123  I looked at the html again and still had the issue with the odd indentation.  Looking at the source, I noticed you used 4 spaces (or tab) to indent.  The docs use 3 spaces (and no tabs) as the standard with an 80 character limit.

Here's the reference in the dev guide for formatting the docs:
https://docs.python.org/devguide/documenting.html#style-guide

oz123 committed:
Format white space to comply with the style guide

oz123 commented:
@csabella fixed the white space. Can you check now?

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
       output (defaulting to the system default encoding).
 
+   :class:`HTMLCalendar` has the following attributes you can override to
+   customize the style of your calender:
```
calender -> calendar

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+
+         cssclasses = ["mon text-bold", "tue", "wed", "thu", "fri", "sat", "sun red"]
+
+      Note that the length of this list must be 7 items.
```
Other places in this page use 'seven' instead of '7'.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+
+   .. attribute:: noday_classes
+
+      The CSS class for a week day occurring in previous or the coming month.
```
- week day -> weekday
- try "in the previous or coming month"

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+   .. attribute:: weekday_head_classes
+
+      A list of CSS styles used for each weekday. The default style list is
+      the same as ``HTMLCalendar.cssclasses``.
```
If you change this to :attr:`HTMLCalendar.cssclasses`, I think Sphinx will make a link.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+   .. attribute:: year_classes
+
+      A space separated list of styles for the table when formatting the year
+      as a table of tables (see: ``HTMLCalendar.formatyear``).
```
If you change this to :meth:`HTMLCalendar.formatyear`, I think Sphinx will make a link.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+   .. attribute:: year_head_classes
+
+      A space separated list of styles for the table head when formatting the
+      year as a table of tables (see: ``HTMLCalendar.formatyear``).
```
If you change this to :meth:`HTMLCalendar.formatyear`, I think Sphinx will make a link.

csabella commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+      year as a table of tables (see: ``HTMLCalendar.formatyear``).
+      The default value is ``"year"``.
+
+Here is an example how you can customize ``HTMLCalendar``::
```
If you use :class:`HTMLCalendar` here, Sphinx will create a link in the doc.

csabella commented:
@oz123 Yes, that fixed the whitespace.  You may want to indent the example at the bottom of the section to line up with the rest.  @doerwalter can comment on this, but you may need to reword lines that have 'you' or 'your' in them.

oz123 committed:
Rewording and addition of sphinx directives

 * change `you` and `your` to more 3 person
 * added sphnix directives instead of unformatted text.

doerwalter commented:
I  agree that the unqualified name `cssclasses` isn't perfect since now there are more `cssclasses` attributes. But for backwards compatibility reasons this should not be changed.

Furthermore the name of the attributes is `...classes`, but the description talks about "styles". IMHO this should be changed from "styles" to "CSS class names".

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+
+   .. attribute:: weekday_head_classes
+
+      A list of CSS styles used for each weekday. The default style list is
```
It is not exactly clear what this refers to. It should probably be something like: "A list of CSS classes names used for the weekday names in the header row about the whole month" or something like that.

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+
+   .. attribute:: month_head_classes
+
+      A space separated list of styles for the month head, for example::
```
-> "A space separated list of CSS class names used for the name of the month above the month table, for example"

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+
+   .. attribute:: month_classes
+
+      A space separated list of styles for the whole table.
```
-> "A space separated list of CSS class names used for the whole month table."

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -170,6 +170,64 @@ it's the base calendar for all computations.
+
+   .. attribute:: year_classes
+
+      A space separated list of styles for the table when formatting the year
```
-> "A space separated list of CSS class names used for the year label above the year table" or something like that.

doerwalter commented:
Also I wonder whether the user might want to use different CSS classes depending on which year or month this is. This would mean that instead of class attributes we would use methods that get passed the year or month or weekday.

doerwalter commented:
With methods we could leave the `cssclasses` attribute undocumented and instead document the method (named `cssclass_day_data` or something like that), which might look like this

```
def cssclass_day_data(self, day, weekday):
   return self.cssclasses[weekday]
```

Other methods would be:

```
def cssclass_weekday_head(self, weekday):
   return self.cssclasses[weekday]

def cssclass_month_data(self, month):
   return "month"

def cssclass_month_head(self, month):
   return "month"

def cssclass_year_data(self, year):
   return "year"

def cssclass_year_head(self, year):
   return "year"
```

Then the user could do something like:

```
def cssclass_month_data(self, month):
   return "month month-%s" % ["", "jan", "feb", "mar", "apr", "may", "jun", "jul", "sep", "oct", "nov", "dec"][month]
```

oz123 commented:
@doerwalter IMHO your suggestion with methods, is pretty complicated. This allows the user to override many methods (which can also override with the old HTMLCalender )
I would prefer simply adding the class attributes, we could add a warning about that old attribute:

```
class HTMLCalendar:
      weekday_classes = ["mon", "tue", "wed", "thu", "fri", "sat", "sun"]
      noday_classes = "noday"
      weekday_head_classes = cssclasses
      month_head_classes = "month"
      month_classes = "month"
      year_head_classes = "year"
      year_classes = "year"
     
      @property
      def cssclasses(self)
            import warnings
            warn("deprecated, please use HTMLCalendar.weekday_classes", DeprecationWarning)
            return self.weekday_classes

...
```

Also, if you feel that `styles` is better we can rename all the attribute from `xxx_class[es]` to
`xxx_style[s]`

doerwalter commented:
Granted, customizing the CSS classes via methods is probably overkill and YAGNI, so we can go with the class attribute approach. I just think, since this is the first time this API gets documented that we should get it right. However if and when this functionality is needed we can extend it without any problems (i.e. these new methods would simply return the appropriate class attribute). But we don't have to do that now.

As for the names: No I think used `cssclasses` is better that using `styles`. The documentation should be updated to be consistent with that and use "CSS classes" everywhere. Also IMHO it would be better to make it clear that `cssclasses` is a list of classes and the other aren't by making the name of the others singular, i.e. use `cssclass_noday`, `cssclass_month` etc.

oz123 commented:
@doerwalter ,

> Also IMHO it would be better to make it clear that cssclasses is a list of classes and the other aren't by making the name of the others singular, i.e. use cssclass_noday, cssclass_month etc.

This is only for the default case, where we supply just one class. I chose to name this `classes` to indicate you can customize this and add your own.

I pushed one more commit implementing the changes you suggested. Hopefully we got it right this time!

Commit:
Refactor Styles -> CSS Classes

Also re-word again the name of the class attributes.
Instead of "classe" use "class" where the attribute is not
a list.
Add a note how to override the attributes easily.

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -175,7 +175,7 @@ it's the base calendar for all computations.
    .. attribute:: cssclasses
 
-      A list of CSS styles used for each weekday. The default style list is::
+      A list of CSS classes used for each weekday. The default style list is::
```
This should read "The default class list is"

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -185,48 +185,51 @@ it's the base calendar for all computations.
-      A list of CSS styles used for each weekday. The default style list is
-      the same as :attr:`cssclasses`.
+      A list of CSS classes used for weekday names in the header row.
+      The default the same as :attr:`cssclasses`.
```
There's an "is" missing in the last sentence

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -185,48 +185,51 @@ it's the base calendar for all computations.
-      A space separated list of styles for the table head when formatting the
-      year as a table of tables (see: :meth:`formatyear`).
+      The CSS class for the whole year's table of tables (see: :meth:`formatyear`).
       The default value is ``"year"``.
```
This and the previous description are the same. They should be distinct, so that the user can determine for whihc part of the page the class gets used.

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -185,48 +185,51 @@ it's the base calendar for all computations.
+      The CSS class for the whole year's table of tables (see: :meth:`formatyear`).
       The default value is ``"year"``.
 
+Note that although the the naming for the above described class attributes is
```
There's a duplicate "the" here.

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -185,48 +185,51 @@ it's the base calendar for all computations.
       The default value is ``"year"``.
 
+Note that although the the naming for the above described class attributes is
+singular (e.g. ``month_class`` ``noday_class``, one can replace the single CSS
```
There's a closing bracket missing here (after noday_class)

doerwalter commented:
I added a few comments to the changes for the documentation. Also IMHO the attributes should be named `cssclass_noday` instead of `noday_class`. This avoids confusion, since it makes it clear, that these are CSS classes, not Python classes. Furthermore by putting the `cssclass` in front it makes it clear that those attributes belong to the same "group" of class attributes.

Sorry for the nitpicking, but I think we're nearly there.

doerwalter commented:
Also, could you take a look at the failing tests?

oz123 committed:
Rename class atrribute again and fix test

 * Prefix class attribute with cssclass_
 This avoids confusion, since it makes it clear, that these are CSS classes,
 not Python classes. Furthermore by putting the cssclass in front it makes it
 clear that those attributes belong to the same "group" of class attributes.

oz123 commented:
@doerwalter  Thanks for all your comments until now! I am sorry I could not reply earlier. I hope this is the last commit, and this will be finally merged.

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -185,39 +185,36 @@ it's the base calendar for all computations.
-singular (e.g. ``month_class`` ``noday_class``, one can replace the single CSS
+Note that although the naming for the above described class attributes is
+singular (e.g. ``cssclass_month`` ``cssclass_noday``), one can replace the single CSS
 class with a space separated list of CSS classes, for example::
```
The line length here is wrong. Could you rewrap this to fit within 80 characters?

doerwalter commented on Lib/calendar.py:
```diff
@@ -381,20 +381,20 @@ class HTMLCalendar(Calendar):
+    cssclass_month = "month"
+    cssclass_year_head = "year"
+    cssclass_year = "year"
 
```
It would be good, if each attribute had a one line comment above it.

doerwalter commented:
Could you add a NEWS entry (in `Misc/NEWS`) for your chances? This should go into the `Library` section of `What's New in Python 3.7.0 alpha 1?`

doerwalter commented:
Also, add yourself to `Misc/ACKS`

doerwalter commented:
Each attribute documentation in `calendar.rst` needs a `.. versionadded:: 3.7`

doerwalter commented:
Also you might want to add an appropriate description of the new features to `Doc/whatsnew/3.7.rst`.

Commit:
Fix wrong usage of CSS classes

Commit:
Add documation strings for the HTMLCalendar class attributes

Commit:
Mention in which version class attributes were added

Commit:
Fix failing test

Commit:
Update NEWS and ACKS

oz123 commented:
@doerwalter I added the changes you suggested.
I really hope these additions to HTMLCalendar will be useful for others too.
Thank you for all the support and suggestions until now.

Commit:
Merge upstream changes for what is new in 3.7

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -185,37 +185,42 @@ it's the base calendar for all computations.
 
       Note that the length of this list must be seven items.
 
+   .. versionadded:: 3.7
```
The `.. versionadded:: 3.7` should go *inside* the `.. attribute::` block (after the description), i.e.

```
    .. attribute:: cssclass_noday
  
        The CSS class for a weekday occurring in the previous or coming month.

        .. versionadded:: 3.7
```

Note that you can create the documentation yourself to check the result via `make html` in the `Doc` directory (you need Sphinx for that)

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -185,37 +185,42 @@ it's the base calendar for all computations.
+   .. versionadded:: 3.7
    .. attribute:: cssclass_month_head
 
       The month's head CSS class. The default value is ``"month"``.
```
You could add a reference to `formatmonthname` here, i.e.:

```
The month's head CSS class (see :meth:`formatmonthname`). The default value is ``"month"``.
```

doerwalter commented on Doc/library/calendar.rst:
```diff
@@ -185,37 +185,42 @@ it's the base calendar for all computations.
+   .. versionadded:: 3.7
    .. attribute:: cssclass_month
 
       The CSS class for the whole month's table. The default value is ``"month"``.
```
You could add a reference to `formatmonth` here, i.e.:

```
The CSS class for the whole month's table (see :meth:`formatmonth`). The default value is ``"month"``.
```

doerwalter commented on Doc/whatsnew/3.7.rst:
```diff
@@ -95,6 +95,13 @@ New Modules
+customisation the CSS classes in the produced HTML calendar.
+(Contributed by Oz Tiram in :issue:`30095`.)
+
+
```
IMHO this should be something like this:

```
calendar
--------

CSS classes used by :class:`~calendar.HTMLCalendar` can now be customized by
subclassing :class:`~calendar.HTMLCalendar` and overwriting certain class
attributes. (Contributed by Oz Tiram in :issue:`30095`.)
```

doerwalter commented on Lib/test/test_calendar.py:
```diff
@@ -844,5 +850,30 @@ def test__all__(self):
+    def test_formatweek(self):
+        weeks = self.cal.monthdays2calendar(2017,5)
+        self.assertIn('class="wed text-nowrap"', self.cal.formatweek(weeks[0]))
+    def test_format_year(self):
```
There's a newline missing here.

oz123 replied:
Fixed in 45d68c2dd0ef5b85581afbdfd3f89c2213bd648e

oz123 committed:
Fix versionadded directives

oz123 committed:
pep8 white space fixes

oz123 committed:
Merge branch 'master' into bpo-30095

doerwalter committed:
Move HTMLCalendar example into class description block.

doerwalter commented:
I pushed a small documentation fix to your branch.

oz123 commented:
@doerwalter WOW! I didn't know you can push to a pull request! That is cool.

doerwalter commented:
@oz123 That's new for me too: see https://docs.python.org/devguide/gitbootcamp.html

doerwalter committed:
Whitespace.

doerwalter committed:
Update variable name that contains a list of CSS classes to be plural.

doerwalter commented:
I've renamed `cssclass_weekday_head` to `cssclasses_weekday_head`, since this is a list of names.

I've noticed that there's no test that checks for this configuration. Could you add one? And if there are other tests missing add those too.

doerwalter committed:
Avoid link to the start of the class description.

doerwalter committed:
In the attribute description refer to the method that uses the attribute.

oz123 committed:
Complete all tests for HTMLCalendar class attributes

oz123 commented:
@doerwalter I think the change you introduce is good. I also added tests to all the class attributes. Hopefully, this is the last iteration.

doerwalter commented:
OK, your patch is in.

Congratulations and thanks for your contribution to Python.
